### PR TITLE
fix #780 fix Object.assign when undefined value and inextensible

### LINKED ISSUE #780: Object.assign() should copy undefined properties, I think
Hi! It looks like `Object.assign()` should copy undefined properties:

```
console.log(Object.assign({"a":undefined}, {"b":undefined}))
Object { a: undefined, b: undefined }
```

However, the Rhino code is skipping undefined values:

```
          case ConstructorId_assign:
          {
            if (args.length < 1) {
              throw ScriptRuntime.typeError1("msg.incompat.call", "assign");
            }
            Scriptable targetObj = ScriptRuntime.toObject(cx, thisObj, args[0]);
            for (int i = 1; i < args.length; i++) {
              if ((args[i] == null) || Undefined.isUndefined(args[i])) {
                continue;
              }
              Scriptable sourceObj = ScriptRuntime.toObject(cx, thisObj, args[i]);
              Object[] ids = sourceObj.getIds();
              for (Object key : ids) {
                if (key instanceof String) {
                  Object val = sourceObj.get((String) key, sourceObj);
                  if ((val != Scriptable.NOT_FOUND) && !Undefined.isUndefined(val)) { // <=== HERE
                    targetObj.put((String) key, targetObj, val);
                  }
                } else if (key instanceof Number) {
                  int ii = ScriptRuntime.toInt32(key);
                  Object val = sourceObj.get(ii, sourceObj);
                  if ((val != Scriptable.NOT_FOUND) && !Undefined.isUndefined(val)) {  // <=== AND HERE
                    targetObj.put(ii, targetObj, val);
                  }
                }
              }
            }
            return targetObj;
          }
```

Is there a reason for those `!Undefined.isUndefined(val)` checks?

Thanks!

### PR DESCRIPTION
**Pull request by tuchida**
ref. #780
```js
Object.keys(Object.assign({a:undefined}, {b:undefined})).join()
// expected:<a[,b]> but was:<a[]>
```
I found another problem with it, so I fixed it while I was at it.
```js
var obj = Object.freeze({});
Object.assign(obj, { a: 1 });
// An error to be thrown, but nothing was thrown.
```

### CODE REVIEW

**Comment by p-bakker** [comment]
Haven't looked at the changes, but haven't these fixes improved the test262 results?

**Comment by tuchida** [comment]
The `undefined` test does not seem to be in test262.
The `Object.freeze` test also does not seem to be in the one used by Rhino. The latest version has it, but it requires the implementation of Computed Property Names.
- https://github.com/tc39/test262/pull/2945

**Comment by p-bakker** [comment]
Right...

> The latest version has it, but it requires the implementation of Computed Property Names.

Not just Computed Property Names though. Also at least a proper const impl and unicode regex

File: src/org/mozilla/javascript/ScriptableObject.java [hunk 1]
```diff
@@ -334,6 +334,38 @@ public void put(Symbol key, Scriptable start, Object value) {
         ensureSymbolScriptable(start).put(key, start, value);
     }
 
+    /**
+     * Set ( O, P, V, Throw = true)
+     *
+     * <p>https://262.ecma-international.org/12.0/#sec-set-o-p-v-throw
+     */
+    public static void putOrThrow(Context cx, Scriptable obj, String name, Object value) {
```
Comment (p-bakker) [inline src/org/mozilla/javascript/ScriptableObject.java:342]: These two Abstract Object Operations ought to be placed in AbstractEcmaObjectOperations. While doing so, check the guidelines at the top of the file

Comment (tuchida) [reply src/org/mozilla/javascript/ScriptableObject.java]: ref. [204cd5f](https://github.com/mozilla/rhino/pull/1186/commits/204cd5fb8ba8f2e5a4c17bf7902fee722cf65260)
I made to move to `AbstractEcmaObjectOperations`.

Comment (tuchida) [reply src/org/mozilla/javascript/ScriptableObject.java]: I rebased it.

**Comment by gbrail** [comment]
Yes -- this looks good. Thanks!

**Pull request merged**